02-linked-lists/02-linked-lists-implementation.md

```diff
@@ -0,0 +1,1030 @@
+![remove a node #3](images/remove-node-3.png)
+![remove a node #4](images/remove-node-4.png)
+
+```python
+def remove(self, index):
+    if not self.head:
+        return
+
+    if index == 0:
+        self.head = self.head.next
+
+    # Traverse the list until you find the node at index -1
+    current = self.head
+    current_index = 0
+    while current.next and current_index < index - 1:
+        current = current.next
+        current_index += 1
+
+    # If the node exists
+    if current.next:
+        current.next = current.next.next
+```
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: 0cc4fc7c-9900-4e3c-9760-608114390a61
+* title: Reverse a linked list
+<!-- * points: [1] (optional, the number of points for scoring as a checkpoint) -->
+<!-- * topics: [python, pandas] (Checkpoints only, optional the topics for analyzing points) -->
+
+##### !question
+
+We can also tackle reversing a linked list by traversing the list and redirecting pointers as we go along. Write a method `reverse` to reverse the given linked list. The function should not return anything.
+
+##### !end-question
+
+##### !placeholder
+
+```py
+class Node:
+    def __init__(self, value):
+        self.val = value
+        self.next = None
+
+class LinkedList:
+    def __init__(self):
+        self.head = None
+
+    def reverse(self, value):
+        #write your code here
+        pass
+```
+
+##### !end-placeholder
+
+##### !tests
+
+[the unit tests below will run against the student submission]
+```py
+import unittest
+from main import *
+
+class LinkedListExtended(LinkedList):
+    
+    def add_first(self, value):
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node
+    
+    def get_first(self):
+        if self.head:
+            return self.head.value
+        return None
+
+    def get_at_index(self, index):
+        current_index = 0
+        current = self.head
+
+        while current_index < index and current:
+            current_index += 1
+            current = current.next
+
+        if current_index == index and current:
+            return current.val
+        return None
+
+
+class TestPython1(unittest.TestCase):
+
+    def test_reverse_empty_list(self):
+        lst = LinkedListExtended()
+
+        self.assertEqual(None, self.get_at_index(0))
+    
+    def test_reverse_will_reverse_five_element_list(self):
+        lst = LinkedListExtended()
+        for i in range(0, 5):
+            lst.add_first(i)
+        
+        lst.reverse()
+
+        for i in range(0,5):
+            self.assertEqual(i, lst.get_at_index(i))
+```
+
+##### !end-tests
+
+##### !hint
+Initialize three pointers: `previous`, `current`, and `next`. `previous` and `next` should be initialized to `Null`, while `current` should start as the `head`.
```

Comment:
It might be helpful to include some kind of gif or short video to demonstrate why/how this works. My brain short-circuits when I try to do it all in my head lol